### PR TITLE
Display attachment duplicate warning

### PR DESCRIPTION
**Pull request by PenghaiZhang**
Based on the discussion, new changes include:

* display the warning message immediately after uploading attachments if duplicates are found
* Clicking this message will redirect users to the Duplicate Data page
* remove the 'View duplicates' button and relevant message from the Save dialog

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
 
 <div class="input text">
 	<#if c.size2 lt 2>
 		<@textfield section=s.field maxlength=1000/>
 	<#else>
 		<@textarea section=s.field rows=c.size2 class="ctrlEditboxMultiLine"/>
 	</#if>
-</div>
\ No newline at end of file
+</div>
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl]: IntelliJ automatically add a new line at the end, but anyway this file will have some changes when I work on adding duplicate warning message to edit boxes.

File: Source/Plugins/Core/com.equella.core/resources/view/wizard/duplicateData.ftl [hunk 2]
```diff
@@ -4,28 +4,47 @@
...
+        <label>${b.key("duplicatedatasection.usedby", [duplicateData.value])}</label>
+        <ul class="blue">
+          <#list duplicateData.items as item>
+            <li><@render item.link/></li>
+          </#list>
+        </ul>
+    </#list>
+  </#if>
+
+</div>
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/resources/view/wizard/duplicateData.ftl]: This file has been reviewed in the previous PR and there are no new changes

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala [hunk 3]
```diff
@@ -168,6 +186,11 @@ class UniversalWebControlNew extends AbstractWebControl[UniversalWebControlModel
 //      sys.error("metadatamapping")
 
       def getId = id
+
+      // Property indicates if duplicate message should be displayed or not
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala]: Used the wrong words. Shouldn't be property... I will write a new comment here.  The two new methods are used in universalattachmentlist.ftl

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala [hunk 4]
```diff
@@ -242,6 +266,26 @@ class UniversalWebControlNew extends AbstractWebControl[UniversalWebControlModel
...
+          for (fileAttachment <- fileAttachments) {
+            wizardService.checkDuplicateAttachments(state,
+                                                    fileAttachment.getFilename,
+                                                    fileAttachment.getUuid)
+          }
+        }
+      }
+      if (attachments.exists(attachment => state.getDuplicateData.containsKey(attachment.getUuid))) {
+        setDuplicateWarning(true)
+      }
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala]: This bit of code has been reviewed in previous PR

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/handlers/FileUploadHandlerNew.scala [hunk 5]
```diff
@@ -204,8 +204,15 @@ class FileUploadHandlerNew extends AbstractAttachmentHandler[FileUploadHandlerMo
...
     da.deleteFiles(ctx.stagingContext)
+    // Remove attachment duplicate information, too
+    if (attachment.isInstanceOf[FileAttachment] && isDuplicateCheck) {
+      ctx.repo.getState.getDuplicateData.remove(attachment.getUuid)
+    }
   }
 
+  def isDuplicateCheck =
+    ctx.controlState.getControlConfiguration.getBooleanAttribute("FILE_DUPLICATION_CHECK")
+
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/handlers/FileUploadHandlerNew.scala]: Changes in this file have been reviewed in previous PR

File: autotest/Tests/src/main/java/com/tle/webtests/pageobject/wizard/DuplicatesTab.java [hunk 6]
```diff
@@ -12,7 +12,7 @@ public void checkWarning(String data) {
     driver
         .findElement(
             By.xpath(
-                "//div[./label[normalize-space(text())="
+                "//div/label[./span[normalize-space(text())="
```
Comment (PenghaiZhang) [inline autotest/Tests/src/main/java/com/tle/webtests/pageobject/wizard/DuplicatesTab.java]: This has been reviewed.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/model/DuplicateData.java [hunk 7]
```diff
@@ -35,9 +35,15 @@
 
   private boolean visible;
   private boolean accepted;
+  // Add a new property to indicate this instance includes attachment duplicate information
+  private boolean attachmentDupCheck;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/model/DuplicateData.java]: Changes in this file have been reviewed.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/DuplicateDataSection.java [hunk 8]
```diff
@@ -177,20 +182,32 @@ public void setupShowingTab(SectionInfo info, SectionTab tab) {
   }
 
   public static class Model {
-    private Collection<DuplicateDataView> duplicateData;
+    private Collection<DuplicateDataView> textfieldDuplicateData;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/DuplicateDataSection.java]: Changes in this class have been reviewed.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/impl/WizardServiceImpl.java [hunk 9]
```diff
@@ -64,6 +64,7 @@
 import com.tle.core.guice.Bind;
 import com.tle.core.guice.BindFactory;
 import com.tle.core.institution.InstitutionService;
+import com.tle.core.item.dao.AttachmentDao;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/impl/WizardServiceImpl.java]: Changes in this file have been reviewed.

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/controls/AbstractWebControl.java [hunk 10]
```diff
@@ -168,6 +170,7 @@ public void evaluate() {
   public void registered(String id, SectionTree tree) {
     this.sectionId = id;
     this.tree = tree;
+    // dupLink.setClickHandler(new OverrideHandler(Js.alert_s("12122")));
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/controls/AbstractWebControl.java]: I will remove these testing code in next commit

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/WizardService.java [hunk 11]
```diff
@@ -83,6 +83,8 @@ boolean checkDuplicateXpathValue(
 
   void checkDuplicateUrls(WizardState state, String[] urls);
 
+  void checkDuplicateAttachments(WizardState state, String fileName, String fileUuid);
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/WizardService.java]: This has been reviewed.

File: Source/Plugins/Core/com.equella.core/resources/view/wizard/duplicateData.ftl [hunk 12]
```diff
@@ -4,28 +4,47 @@
...
-						<li><@render item.link/></li>
-					</#list>
-				</ul>
-			</div>
-		</#if>
-	</#list>
-</div>
\ No newline at end of file
+
+  <#if m.textfieldDuplicateData?size gt 0>
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/resources/view/wizard/duplicateData.ftl]: This file has been reviewed in the previous PR and there are no new changes

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala [hunk 13]
```diff
@@ -242,6 +266,26 @@ class UniversalWebControlNew extends AbstractWebControl[UniversalWebControlModel
           )
         )
       }
+
+      val attachments = dialog.getAttachments.asScala
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/controls/universal/UniversalWebControlNew.scala]: This bit of code has been reviewed in previous PR

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/command/SaveDialog.java [hunk 14]
```diff
@@ -73,7 +73,7 @@
   private static Label LABEL_PUBLISH_NOMOD_MSG;
 
   @PlugKey("command.save.successreceipt")
-  private static Label SUCCESS_RECEIPT_LABEL;
+  private static Label LABEL_SUCCESS_RECEIPT;
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/command/SaveDialog.java]: This  has been reviewed in previous PR

File: Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/WizardBodySection.java [hunk 15]
```diff
@@ -416,6 +417,25 @@ public boolean goToFirstUnfinished(SectionInfo info) {
...
+    // and it's impossible to have more than 1 of this tab
+    SectionTab duplicateTab =
+        tabs.stream()
+            .filter(tab -> tab.getSectionName().equals(DuplicateDataSection.PROP_NAME))
+            .collect(Collectors.toList())
+            .get(0);
+    int oldTab = model.getCurrentTab();
+    SectionTab currentTab = tabs.get(oldTab);
+    // Must explicitly leave current tab before going another tab
+    currentTab.getTabSection().leavingTab(info, currentTab);
```
Comment (PenghaiZhang) [inline Source/Plugins/Core/com.equella.core/src/com/tle/web/wizard/section/WizardBodySection.java]: Apart from line 430-433, others have been reviewed.

File: Source/Plugins/Core/com.equella.core/resources/lang/i18n-resource-centre.properties [hunk 16]
```diff
@@ -1210,7 +1210,9 @@ duplicatedatasection.mustchange=You <b>must</b> change the values of the fields
 duplicatedatasection.mustchangesymbol=*
 duplicatedatasection.pagename=Duplicate data
 duplicatedatasection.preamble=The following data is duplicated by other items in this collection.
+duplicatedatasection.preambleforattachment=The following data are duplicates of your attachments in this collection.
 duplicatedatasection.usedby=''{0}'' is also used by\:
+duplicatewarningmessage=Duplicates exist in other items. View duplicate details here
```
[the remark this answers is not in the document]
Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/resources/lang/i18n-resource-centre.properties]: Oh, when we discussed this the 'duplicates here' bit was meant to be a link. (And then it would just end with a full stop.)

But I guess you could end with a colon and then have a button.

File: Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl [hunk 17]
```diff
@@ -1,11 +1,12 @@
 <#include "/com.tle.web.freemarker@/macro/sections.ftl"/>
 <#include "/com.tle.web.sections.standard@/textarea.ftl"/>
 <#include "/com.tle.web.sections.standard@/textfield.ftl"/>
+<#include "/com.tle.web.sections.standard@/link.ftl">
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl]: Ummm, why is this included here if not used below?

Indeed, why was this file modified at all?

File: Source/Plugins/Core/com.equella.core/resources/view/wizard/savedialog.ftl [hunk 18]
```diff
@@ -13,4 +13,4 @@
 			<@textarea section=s.message class="focus"/>
 		</div>
 	</#if>
-</div>
\ No newline at end of file
+</div>
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/resources/view/wizard/savedialog.ftl]: You should only commit files that contain changes you need. This is a matter of ensuring your review your diffs before you commit files.

File: Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl (hunk 17 above)
Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl]: Ahh sorry I forgot to remove this include. 
When working on file duplicate checking I also found this file is where will make changes for the edit box duplicate check. So I did some local experiments and succeeded, but forgot to remove this line. I will remove it in next commit

Comment (PenghaiZhang) [reply Source/Plugins/Core/com.equella.core/resources/view/editbox.ftl]: Ahh sorry I forgot to remove this include.
When working on file duplicate checking I also found this file is one of those files that I will make changes for the edit box duplicate check. So I did some local experiments and succeeded, but forgot to remove this line. I will remove it in next commit.
